engine/execution/computation/computer/transaction_coordinator.go

```diff
@@ -0,0 +1,185 @@
+	writeBehindLog TransactionWriteBehindLogger,
+) *transactionCoordinator {
+	mutex := &sync.Mutex{}
+	cond := sync.NewCond(mutex)
+
+	database := storage.NewBlockDatabase(
+		storageSnapshot,
+		0,
+		cachedDerivedBlockData)
+
+	return &transactionCoordinator{
+		vm:             vm,
+		mutex:          mutex,
+		cond:           cond,
+		snapshotTime:   0,
+		abortErr:       nil,
+		database:       database,
+		writeBehindLog: writeBehindLog,
+	}
+}
+
+func (coordinator *transactionCoordinator) SnapshotTime() logical.Time {
+	coordinator.mutex.Lock()
+	defer coordinator.mutex.Unlock()
+
+	return coordinator.snapshotTime
+}
+
+func (coordinator *transactionCoordinator) Error() error {
+	coordinator.mutex.Lock()
+	defer coordinator.mutex.Unlock()
+
+	return coordinator.abortErr
+}
+
+func (coordinator *transactionCoordinator) AbortAllOutstandingTransactions(
+	err error,
+) {
+	coordinator.mutex.Lock()
+	defer coordinator.mutex.Unlock()
+
+	if coordinator.abortErr != nil { // Transactions are already aborting.
+		return
+	}
+
+	coordinator.abortErr = err
+	coordinator.cond.Broadcast()
+}
+
+func (coordinator *transactionCoordinator) NewTransaction(
+	request TransactionRequest,
+) (
+	*transaction,
+	error,
+) {
+	err := coordinator.Error()
+	if err != nil {
+		return nil, err
+	}
+
+	txn, err := coordinator.database.NewTransaction(
+		request.ExecutionTime(),
+		fvm.ProcedureStateParameters(request.ctx, request))
+	if err != nil {
+		return nil, err
+	}
+
+	return &transaction{
+		request:     request,
+		coordinator: coordinator,
+		startedAt:   time.Now(),
+		Transaction: txn,
+		ProcedureExecutor: coordinator.vm.NewExecutor(
+			request.ctx,
+			request.TransactionProcedure,
+			txn),
+	}, nil
+}
+
+func (coordinator *transactionCoordinator) commit(txn *transaction) error {
+	coordinator.mutex.Lock()
+	defer coordinator.mutex.Unlock()
+
+	if coordinator.abortErr != nil {
+		return coordinator.abortErr
+	}
+
+	executionSnapshot, err := txn.Transaction.Commit()
+	if err != nil {
+		return err
+	}
+
+	coordinator.writeBehindLog.AddTransactionResult(
+		txn.request,
+		executionSnapshot,
+		txn.Output(),
+		time.Since(txn.startedAt))
+
+	coordinator.snapshotTime += 1
+	coordinator.cond.Broadcast()
+
+	return nil
+}
+
+func (txn *transaction) Commit() error {
+	return txn.coordinator.commit(txn)
+}
+
+func (coordinator *transactionCoordinator) waitForUpdatesNewerThan(
+	snapshotTime logical.Time,
+) (
+	logical.Time,
+	error,
+	logical.Time,
+	error,
+) {
+	coordinator.mutex.Lock()
+	defer coordinator.mutex.Unlock()
+
+	startTime := coordinator.snapshotTime
+	startErr := coordinator.abortErr
+	for coordinator.snapshotTime <= snapshotTime && coordinator.abortErr == nil {
+		coordinator.cond.Wait()
+	}
+
+	return startTime, startErr, coordinator.snapshotTime, coordinator.abortErr
+}
+
+func (txn *transaction) WaitForUpdates() error {
+	_, _, _, err := txn.coordinator.waitForUpdatesNewerThan(txn.SnapshotTime())
```

Comment:
why does `waitForUpdatesNewerThan` need to return two pairs of (value, error), why the `startErr` is being ignored?